### PR TITLE
Remove graph breaks for torch.compile() in flash_attention_forward when Lllama Model is padding free tuned

### PR DESCRIPTION
**Pull request by Abhishek-TAMU**
# What does this PR do?

This PR removes the function call `prepare_fa2_from_position_ids` in `flash_attention_forward` as it causes graph break when `torch_compile` flag is turned on in Training [arguments](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.TrainingArguments.torch_compile) to use in SFTTrainer to perform padding free tuning of Llama model. This is because code in `prepare_fa2_from_position_ids` incur a cpu-gpu sync that is unavoidable. 
Hence `cu_seq_lens_q`, `cu_seq_lens_k`, `max_length_k`, `max_length_q` is now taken from the batch in `DataCollatorForCompletionOnlyLM` with this [PR](https://github.com/huggingface/trl/pull/2158) to avoid call to `prepare_fa2_from_position_ids` in `flash_attention_forward`. 

CC: @ani300 @ArthurZucker

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -953,6 +979,10 @@ def forward(
                     use_cache=use_cache,
                     cache_position=cache_position,
                     position_embeddings=position_embeddings,
+                    cu_seq_lens_q=cu_seq_lens_q,
+                    cu_seq_lens_k=cu_seq_lens_k,
+                    max_length_q=max_length_q,
+                    max_length_k=max_length_k,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:985]: Actually something we had planned 😅 cc @gante on generate unpadding the input!

Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:985]: @Cyrilvallez as well if you want to have fun IMO can be quite impactfull!

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1148,6 +1178,10 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
+        cu_seq_lens_q: Optional[torch.LongTensor] = None,
+        cu_seq_lens_k: Optional[torch.LongTensor] = None,
+        max_length_q: int = 0,
+        max_length_k: int = 0,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1184]: these are are FlashAttention specific. IMO it would make sense to just add them as `fa2_kwargs` for example. We can use something like this: https://github.com/huggingface/transformers/blob/ce81eb8d14ae39e183a8956cce3f7e2e46ecaa8f/src/transformers/processing_utils.py#L81

**Review by ArthurZucker — COMMENTED** [review]
A very nice PR and very much welcome! 
Let's add general kwargs, #31446 has some commits with that 
image

**Comment by ArthurZucker** [comment]
This way we can potentially add more kwargs without changing the forward!

**Comment by Abhishek-TAMU** [comment]
Thanks for the review and welcoming this PR.
The changes suggested by you have been made @ArthurZucker

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -422,10 +425,7 @@ def forward(
         use_cache: bool = False,
         cache_position: Optional[torch.LongTensor] = None,
         position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # will become mandatory in v4.46
-        cu_seq_lens_q: Optional[torch.LongTensor] = None,
-        cu_seq_lens_k: Optional[torch.LongTensor] = None,
-        max_length_q: int = 0,
-        max_length_k: int = 0,
+        **kwargs,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:427]:
```suggestion
        **kwargs: Unpack[Fa2Kwargs],
```

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -891,11 +880,7 @@ def forward(
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
-        cache_position: Optional[torch.LongTensor] = None,
-        cu_seq_lens_q: Optional[torch.LongTensor] = None,
-        cu_seq_lens_k: Optional[torch.LongTensor] = None,
-        max_length_q: int = 0,
-        max_length_k: int = 0,
+        **fa2_kwargs: Fa2Kwargs,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:882]:
```suggestion
        **fa2_kwargs: Unpack[Fa2Kwargs],
```

**Review by ArthurZucker — COMMENTED** [review]
Nice!

**Review by ArthurZucker — COMMENTED** [review]
Thinking that we can call then `flash_attn_kwargs` to not depend on versioning !

**Comment by Abhishek-TAMU** [comment]
@ArthurZucker Made the necessary changes. 
Feel free to suggest changes if required any. Thanks!

**Comment by Abhishek-TAMU** [comment]
@ArthurZucker Would you mind facilitating in moving ahead with the related PR in TRL which supports this PR: https://github.com/huggingface/trl/pull/2158 ?

**Comment by ArthurZucker** [comment]
Okay! Overall looks good. 
1. we need to protect the import of Unpack : 
2. let's just add an example in the documentation of how to use this! a small python snippet!

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -18,7 +18,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import math
-from typing import List, Optional, Tuple, Union
+from typing import List, Optional, Tuple, Union, Unpack
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:21]:
```suggestion
from typing import List, Optional, Tuple, Union
```

File: src/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -48,6 +48,9 @@
     logging,
     replace_return_docstrings,
 )
+from ...processing_utils import (
+    FlashAttentionKwargs,
+)
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:53]:
```suggestion
from ...processing_utils import (
    FlashAttentionKwargs, Unpack
)    
```

**Comment by ArthurZucker** [comment]
The CI's should mostly go green with this! 
Then you will have `make fix-copies` that is gonna propagate the changes!

**Comment by ArthurZucker** [comment]
Once we merge I'll ping TRL team to make sure they don't miss it!

**Comment by Abhishek-TAMU** [comment]
@ArthurZucker Changes have been pushed to fix CI tests.

**Comment by Abhishek-TAMU** [comment]
For documentation does below sample script works which only performs tuning of Llama model using `SFTTrainer` with `torch_compile` flag on and using DataCollator with padding free ? @ArthurZucker 


```
#################### IMPORTS ###################

import math
import datasets, dataclasses
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    TrainingArguments
)
from trl import SFTConfig, SFTTrainer, DataCollatorForCompletionOnlyLM

#################### MODEL LOADING USING FLASH ATTENTION ###################

model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B", attn_implementation="flash_attention_2")
tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)

#################### DATA PREPROCESSING (PADDING_FREE) ###################

response_template = "\n### Label:"
response_template_ids = tokenizer.encode(
    response_template, add_special_tokens=False
)[2:]
data_collator = DataCollatorForCompletionOnlyLM(
    response_template_ids,
    tokenizer=tokenizer,
    ignore_index= -100,
    padding_free=True
)
format_dataset = lambda example: {  
    "output": example["output"] + tokenizer.eos_token
}
data_files = {"train": "path/to/dataset"}
json_dataset = datasets.load_dataset("json", data_files=data_files)
formatted_train_dataset = json_dataset["train"].map(format_dataset)

################# Trainer CONFIG ############################

train_args = TrainingArguments(
    num_train_epochs=5,
    per_device_train_batch_size=4,
    per_device_eval_batch_size=4,
    gradient_accumulation_steps=4,
    learning_rate=0.00001,
    weight_decay=0,
    warmup_ratio=0.03,
    lr_scheduler_type="cosine",
    logging_steps=1,
    include_tokens_per_second=True,
    save_strategy="epoch",
    output_dir="tmp",
    torch_compile=True,
    torch_compile_backend="inductor",
    torch_compile_mode="default"
)
transformer_train_arg_fields = [x.name for x in dataclasses.fields(SFTConfig)]
transformer_kwargs = {
    k: v
    for k, v in train_args.to_dict().items()
    if k in transformer_train_arg_fields
}
training_args = SFTConfig(**transformer_kwargs)

####################### TUNING #####################

trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=formatted_train_dataset,
    data_collator=data_collator,
    dataset_text_field="output",
    args=training_args,
)
trainer.train()
```

**Comment by ArthurZucker** [comment]
Yep looks perfect! Especially if we can have % improvement !

File: src/transformers/modeling_flash_attention_utils.py [hunk 7]
```diff
@@ -267,24 +276,15 @@ def _flash_attention_forward(
...
     # Use `flash_attn_varlen_func` to prevent cross-example attention and also allow padding free approach
-    # Note: the `torch.diff(...)` condition is last to use short-circuit and avoid the cuda synchronization it incurs during inference (query_length == 1 always)
-    elif position_ids is not None and query_length != 1 and not (torch.diff(position_ids, dim=-1) >= 0).all():
-        batch_size = query_states.size(0)
-        query_states, key_states, value_states, indices_q, cu_seq_lens, max_seq_lens = prepare_fa2_from_position_ids(
-            query_states, key_states, value_states, position_ids
-        )
-
-        cu_seqlens_q, cu_seqlens_k = cu_seq_lens
-        max_seqlen_in_batch_q, max_seqlen_in_batch_k = max_seq_lens
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:278]: We should keep this in case cu_seq_lens_q/k and etc are not passsed to compute them!

File: src/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -494,6 +496,11 @@ def forward(
             key_states = key_states.to(target_dtype)
             value_states = value_states.to(target_dtype)
 
+        batch_size = query_states.size(0)
+        query_states = query_states.reshape(-1, query_states.size(-2), query_states.size(-1))
+        key_states = key_states.reshape(-1, key_states.size(-2), key_states.size(-1))
+        value_states = value_states.reshape(-1, value_states.size(-2), value_states.size(-1))
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:502]:
```suggestion
```
I think we should do this in the `_flash_attention_forward` wrapper this we have 0 modeling changes, and all models will benefit easily from this!

**Review by ArthurZucker — APPROVED** [review]
🚀 last few nits and we can add these flash_attn_kwargs to all models afterwards! Great contribution

**Comment by Abhishek-TAMU** [comment]
Made changes as suggested. Feel free to have a look at overall changes to confirm @ArthurZucker . Also which section can I add the above documentation script ?

**Comment by ArthurZucker** [comment]
Probably under this section: https://huggingface.co/docs/transformers/llm_optims#flashattention-2

File: src/transformers/modeling_flash_attention_utils.py [hunk 9]
```diff
@@ -180,6 +180,10 @@ def prepare_fa2_from_position_ids(query, key, value, position_ids):
     return (query, key, value, indices_q, (cu_seq_lens, cu_seq_lens), (max_length, max_length))
 
 
+flash_241 = is_flash_attn_greater_or_equal("2.4.1")
+deterministic_g = os.environ.get("FLASH_ATTENTION_DETERMINISTIC", "0") == "1"
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:184]: nice catch this was breaking compile on my side as well

**Review by ArthurZucker — APPROVED** [review]
LGTM letting you resolve the conflicts and we can merge!

**Review by ArthurZucker — COMMENTED** [review]
Okay! This LGTM!
You just need to run the `make fix-copies` to make sure CIs go green 🚀

**Comment by ArthurZucker** [comment]
If you have issue with just make fix copies I can take it over if you want!

**Comment by Abhishek-TAMU** [comment]
That would be helpful. Thank you! There seems to be some mismatch in `src/transformers/models/glm/modeling_glm.py`.

**Comment by ArthurZucker** [comment]
The new helper will be this:
image

(with loss kwargs!)

**Comment by Abhishek-TAMU** [comment]
Thanks @ArthurZucker for the code change to accomodate LossKwargs.

**Comment by Abhishek-TAMU** [comment]
> Once we merge I'll ping TRL team to make sure they don't miss it

Hi @ArthurZucker, do you mind facilitating this ?

**Comment by ma787639046** [comment]
Hi @Abhishek-TAMU @ArthurZucker , very nice PR for adding FlashAttentionKwargs. 
In [modeling_llama.py#L959](https://github.com/Abhishek-TAMU/transformers/blob/cb08b6371101dae92867d52b1a3b24405779d46f/src/transformers/models/llama/modeling_llama.py#L959), I noticed that `**flash_attn_kwargs` is added to the inputs of `decoder_layer` when gradient checkpointing is not used. Could you please also add `flash_attn_kwargs` when using gradient checkpointing in the above [if branch at line 938](https://github.com/Abhishek-TAMU/transformers/blob/cb08b6371101dae92867d52b1a3b24405779d46f/src/transformers/models/llama/modeling_llama.py#L938)? If checkpointing function does not accept kwargs, can we make all FlashAttentionKwargs as optional input fields of `LlamaDecoderLayer.forward`?

**Comment by ArthurZucker** [comment]
kwargs are by default optional! 
Checkpoint indeed does not support kwargs, at least the way we have formulated it. Will be fixed by #34987
